File: src/risper/open.py
from __future__ import annotations

import argparse
import sys
from pathlib import Path

from .config import load_config
from .platforms import current_platform
from .session_actions import copy_transcript, transcript_path
from .sessions import last_session


def _open_path(path: Path) -> int:
    ok, message = current_platform().open_path(path)
    if not ok:
        print(message, file=sys.stderr)
        return 1
    return 0


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="Open Risper files and folders.")
    parser.add_argument(
        "target",
        nargs="?",
        default="recordings",
        choices=["recordings", "last-session", "last-transcript", "last-audio", "play-last", "config", "copy-last"],
    )
    args = parser.parse_args(argv)
    config = load_config()

    if args.target == "recordings":
        return _open_path(config.sessions_dir)
    if args.target == "config":
        return _open_path(config.config_path)

    metadata = last_session(config)
    if not metadata:
        print("No Risper sessions yet.", file=sys.stderr)
        return 1
    session_dir = Path(str(metadata["audio_path"])).parent

    if args.target == "last-session":
        return _open_path(session_dir)
    if args.target in {"last-audio", "play-last"}:
        return _open_path(Path(str(metadata["audio_path"])))
    if args.target == "last-transcript":
        transcript = transcript_path(metadata)
        if not transcript:
            print("Last session has no transcript.", file=sys.stderr)
            return 1
        return _open_path(transcript)
    if args.target == "copy-last":
        ok, message = copy_transcript(metadata)
        print(message)
        return 0 if ok else 1
    return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: open/open.go
// Package open implements the files-and-folders command.
package open

import (
	"flag"
	"fmt"
	"os"

	"github.com/rob137/risper/config"
	"github.com/rob137/risper/desktop"
	"github.com/rob137/risper/internal/args"
	"github.com/rob137/risper/session"
	"github.com/rob137/risper/sessionactions"
)

var targets = map[string]bool{
	"recordings": true, "last-session": true, "last-transcript": true,
	"last-audio": true, "play-last": true, "config": true, "copy-last": true,
}

func Main(argv []string) int {
	parser := flag.NewFlagSet("risper open", flag.ContinueOnError)
	parser.SetOutput(os.Stderr)
	if err := parser.Parse(args.Reorder(argv, nil)); err != nil {
		if err == flag.ErrHelp {
			return 0
		}
		return 2
	}
	target := "recordings"
	if parser.NArg() > 1 {
		fmt.Fprintln(os.Stderr, "risper open: unexpected positional arguments")
		return 2
	}
	if parser.NArg() == 1 {
		target = parser.Arg(0)
	}
	if !targets[target] {
		fmt.Fprintf(os.Stderr, "risper open: invalid target %q\n", target)
		return 2
	}
	cfg, err := config.Load()
	if err != nil {
		return reportError("risper open", err)
	}
	if target == "recordings" {
		return openPath(cfg.SessionsDir)
	}
	if target == "config" {
		return openPath(cfg.ConfigPath)
	}
	metadata, err := session.Last(cfg)
	if err != nil {
		return reportError("risper open", err)
	}
	if metadata == nil {
		fmt.Fprintln(os.Stderr, "No Risper sessions yet.")
		return 1
	}
	switch target {
	case "last-session":
		return openPath(session.SessionDir(metadata))
	case "last-audio", "play-last":
		return openPath(metadata.AudioPath)
	case "last-transcript":
		path := sessionactions.TranscriptPath(metadata)
		if path == "" {
			fmt.Fprintln(os.Stderr, "Last session has no transcript.")
			return 1
		}
		return openPath(path)
	case "copy-last":
		ok, message := sessionactions.CopyTranscript(metadata)
		fmt.Println(message)
		if ok {
			return 0
		}
		return 1
	default:
		return 2
	}
}

func openPath(path string) int {
	ok, message := desktop.OpenPath(path)
	if !ok {
		fmt.Fprintln(os.Stderr, message)
		return 1
	}
	return 0
}

func reportError(prefix string, err error) int {
	fmt.Fprintln(os.Stderr, prefix+":", err)
	return 1
}
